ci: fix intermittent failures of the main-tests and pbs-tests test

As observed by @costrouc in https://github.com/dask/dask-gateway/pull/392#issuecomment-829615662 and me in #420, the PBS test is intermittently failing unrelated to changes in PRs.

Let's try fix that!

Also the main test seem intermittent.

![image](https://user-images.githubusercontent.com/3837114/130842037-40b5cdbb-e9d9-4702-b926-b665e1e5d7d4.png)

### Logs from pbs test

```
Run ./continuous_integration/docker/pbs/start.sh
++ dirname ./continuous_integration/docker/pbs/start.sh
+ this_dir=./continuous_integration/docker/pbs
++ cd ./continuous_integration/docker/pbs
++ pwd
+ full_path_this_dir=/home/runner/work/dask-gateway/dask-gateway/continuous_integration/docker/pbs
++ cd /home/runner/work/dask-gateway/dask-gateway/continuous_integration/docker/pbs/../../..
++ pwd
+ git_root=/home/runner/work/dask-gateway/dask-gateway
+ docker run --rm -d --name pbs -h pbs -v /home/runner/work/dask-gateway/dask-gateway:/working -p 8000:8000 -p 8786:8786 -p 8088:8088 --cap-add=SYS_RESOURCE daskgateway/testing-pbs
Unable to find image 'daskgateway/testing-pbs:latest' locally
latest: Pulling from daskgateway/testing-pbs
ac9208207ada: Pulling fs layer
e9270b8e0c2a: Pulling fs layer
404a529e534a: Pulling fs layer
8ebdb67ecf9e: Pulling fs layer
10da784d767c: Pulling fs layer
4c7ef137d5c7: Pulling fs layer
10da784d767c: Waiting
4c7ef137d5c7: Waiting
8ebdb67ecf9e: Waiting
e9270b8e0c2a: Verifying Checksum
e9270b8e0c2a: Download complete
8ebdb67ecf9e: Verifying Checksum
8ebdb67ecf9e: Download complete
ac9208207ada: Verifying Checksum
ac9208207ada: Download complete
4c7ef137d5c7: Verifying Checksum
4c7ef137d5c7: Download complete
404a529e534a: Verifying Checksum
404a529e534a: Download complete
ac9208207ada: Pull complete
10da784d767c: Verifying Checksum
10da784d767c: Download complete
e9270b8e0c2a: Pull complete
404a529e534a: Pull complete
8ebdb67ecf9e: Pull complete
10da784d767c: Pull complete
4c7ef137d5c7: Pull complete
Digest: sha256:50c549b726cc12198ac813a51101dc42f67752d257f765f1b6e774c370c124d4
Status: Downloaded newer image for daskgateway/testing-pbs:latest
f590a3f3fc32c3e0465fa2d7171e2c284c3a03b4cfbc26bddbf262400eca63e6
+ cd /working
+ conda install psutil

dask-gateway/dask_gateway/client.py:581: GatewayClusterError
----------------------------- Captured stderr call -----------------------------
[I 2021-08-25 17:40:57.307 DaskGateway] Starting dask-gateway-server - version 0.9.0
[I 2021-08-25 17:40:57.308 DaskGateway] Authenticator: 'dask_gateway_server.auth.SimpleAuthenticator'
[I 2021-08-25 17:40:57.308 DaskGateway] Backend: 'tests.test_pbs_backend.PBSTestingBackend'
[I 2021-08-25 17:40:57.311 DaskGateway] Generating new api token for proxy
[I 2021-08-25 17:40:57.311 DaskGateway] Starting the Dask gateway proxy...
[I 2021-08-25 17:40:57.314 DaskGateway] Dask gateway proxy started
[I 2021-08-25 17:40:57.314 DaskGateway] - HTTP routes listening at http://127.0.0.1:53383
[I 2021-08-25 17:40:57.314 DaskGateway] - Scheduler routes listening at gateway://127.0.0.1:53383
[W 2021-08-25 17:40:57.316 Proxy] Unexpected failure fetching routing table, retrying in 0.5s: Get http://127.0.0.1:34421/api/v1/routes: dial tcp 127.0.0.1:34421: connect: connection refused
[I 2021-08-25 17:40:57.330 DaskGateway] Backend started, clusters will contact api server at http://127.0.0.1:53383/api
[I 2021-08-25 17:40:57.332 DaskGateway] Dask-Gateway server started
[I 2021-08-25 17:40:57.332 DaskGateway] - Private API server listening at http://127.0.0.1:34421
[I 2021-08-25 17:40:57.819 DaskGateway] 200 GET /api/v1/routes 0.581ms
[W 2021-08-25 17:40:57.823 DaskGateway] 401 POST /api/v1/clusters/ 0.457ms
[I 2021-08-25 17:40:57.909 DaskGateway] Created cluster ee75e379c8d648dfbcfa0755dc6e221d for user alice
[I 2021-08-25 17:40:57.910 DaskGateway] 201 POST /api/v1/clusters/ 86.218ms
[I 2021-08-25 17:40:57.910 DaskGateway] Submitting cluster ee75e379c8d648dfbcfa0755dc6e221d...
[W 2021-08-25 17:40:58.143 DaskGateway] Failed to submit cluster ee75e379c8d648dfbcfa0755dc6e221d
Traceback (most recent call last):
  File "/working/dask-gateway-server/dask_gateway_server/backends/db_base.py", line 1248, in _cluster_to_submitted
    async for state in self.do_start_cluster(cluster):
  File "/working/tests/test_pbs_backend.py", line 52, in do_start_cluster
    async for state in super().do_start_cluster(cluster):
  File "/working/dask-gateway-server/dask_gateway_server/backends/jobqueue/base.py", line 178, in do_start_cluster
    job_id = await self.start_job(
  File "/working/dask-gateway-server/dask_gateway_server/backends/jobqueue/base.py", line 128, in start_job
    raise Exception(
Exception: Failed to submit job to batch system
  exit_code: 189
  stdout: 
  stderr: qsub: No default queue specified

[I 2021-08-25 17:40:58.147 DaskGateway] Stopping cluster ee75e379c8d648dfbcfa0755dc6e221d...
[I 2021-08-25 17:40:58.148 DaskGateway] Cluster ee75e379c8d648dfbcfa0755dc6e221d stopped
[I 2021-08-25 17:40:58.149 DaskGateway] 200 GET /api/v1/clusters/ee75e379c8d648dfbcfa0755dc6e221d?wait 231.451ms
[I 2021-08-25 17:40:58.150 DaskGateway] 204 DELETE /api/v1/clusters/ee75e379c8d648dfbcfa0755dc6e221d 0.283ms
[W 2021-08-25 17:40:58.151 Proxy] Unexpected failure fetching routing events, retrying in 0.5s: Get http://127.0.0.1:34421/api/v1/routes?watch=1&last_id=1: dial tcp 127.0.0.1:34421: connect: connection refused
[I 2021-08-25 17:40:58.152 DaskGateway] Stopping the Dask gateway proxy
[I 2021-08-25 17:40:58.152 DaskGateway] Stopped successfully
=============================== warnings summary ===============================
../opt/miniconda/lib/python3.8/site-packages/_pytest/cacheprovider.py:428
  /opt/miniconda/lib/python3.8/site-packages/_pytest/cacheprovider.py:428: PytestCacheWarning: could not create cache path /working/.pytest_cache/v/cache/nodeids
    config.cache.set("cache/nodeids", sorted(self.cached_nodeids))

../opt/miniconda/lib/python3.8/site-packages/_pytest/cacheprovider.py:382
  /opt/miniconda/lib/python3.8/site-packages/_pytest/cacheprovider.py:382: PytestCacheWarning: could not create cache path /working/.pytest_cache/v/cache/lastfailed
    config.cache.set("cache/lastfailed", self.lastfailed)

../opt/miniconda/lib/python3.8/site-packages/_pytest/stepwise.py:49
  /opt/miniconda/lib/python3.8/site-packages/_pytest/stepwise.py:49: PytestCacheWarning: could not create cache path /working/.pytest_cache/v/cache/stepwise
    session.config.cache.set(STEPWISE_CACHE_DIR, [])

-- Docs: https://docs.pytest.org/en/stable/warnings.html
=========================== short test summary info ============================
FAILED tests/test_pbs_backend.py::test_pbs_backend - dask_gateway.client.Gate...
=================== 1 failed, 1 passed, 3 warnings in 1.45s ====================
```

### Logs from main test

```
Run py.test tests/ -k 'not kubernetes' -v
============================= test session starts ==============================
platform linux -- Python 3.8.11, pytest-6.2.4, py-1.10.0, pluggy-0.13.1 -- /opt/hostedtoolcache/Python/3.8.11/x64/bin/python
cachedir: .pytest_cache
rootdir: /home/runner/work/dask-gateway/dask-gateway
plugins: asyncio-0.12.0
collecting ... collected 116 items / 3 deselected / 5 skipped / 108 selected

tests/test_auth.py::test_basic_auth PASSED                               [  0%]
tests/test_auth.py::test_basic_auth_password PASSED                      [  1%]
tests/test_auth.py::test_kerberos_auth SKIPPED (Cannot import kerberos)  [  2%]
tests/test_auth.py::test_jupyterhub_auth_user PASSED                     [  3%]
tests/test_auth.py::test_jupyterhub_auth_service PASSED                  [  4%]
tests/test_cli.py::test_generate_config PASSED                           [  5%]
tests/test_cli.py::test_proxy_cli PASSED                                 [  6%]
tests/test_client.py::test_get_auth PASSED                               [  7%]
tests/test_client.py::test_config_auth_kwargs_template_environment_vars PASSED [  7%]
tests/test_client.py::test_jupyterhub_auth PASSED                        [  8%]
tests/test_client.py::test_client_init PASSED                            [  9%]
tests/test_client.py::test_gateway_proxy_address_infer_port PASSED       [ 10%]
tests/test_client.py::test_gateway_addresses_template_environment_vars PASSED [ 11%]
tests/test_client.py::test_http_client_proxy_false PASSED                [ 12%]
tests/test_client.py::test_http_client_proxy_true PASSED                 [ 13%]
tests/test_client.py::test_http_client_proxy_explicit PASSED             [ 14%]
tests/test_client.py::test_get_versions PASSED                           [ 15%]
tests/test_client.py::test_client_reprs PASSED                           [ 15%]
tests/test_client.py::test_cluster_widget PASSED                         [ 16%]
tests/test_client.py::test_dashboard_link_from_public_address SKIPPED    [ 17%]
tests/test_client.py::test_create_cluster_with_GatewayCluster_constructor PASSED [ 18%]
tests/test_client.py::test_sync_constructors PASSED                      [ 19%]
tests/test_client.py::test_GatewayCluster_shutdown_on_close PASSED       [ 20%]
tests/test_client.py::test_GatewayCluster_client_error_doesnt_prevent_cleanup PASSED [ 21%]
tests/test_client.py::test_GatewayCluster_cleanup_atexit PASSED          [ 22%]
tests/test_db_backend.py::test_shutdown_on_startup_error PASSED          [ 23%]
tests/test_db_backend.py::test_db_encrypt_keys_required PASSED           [ 23%]
tests/test_db_backend.py::test_db_encrypt_keys_invalid PASSED            [ 24%]
tests/test_db_backend.py::test_db_encrypt_keys_from_env PASSED           [ 25%]
tests/test_db_backend.py::test_resume_clusters_forbid_in_memory_db PASSED [ 26%]
tests/test_db_backend.py::test_encryption PASSED                         [ 27%]
tests/test_db_backend.py::test_normalize_encrypt_key PASSED              [ 28%]
tests/test_db_backend.py::test_cleanup_expired_clusters PASSED           [ 29%]
tests/test_db_backend.py::test_slow_cluster_start[0.1-1] PASSED          [ 30%]
tests/test_db_backend.py::test_slow_cluster_start[0.4-2] PASSED          [ 30%]
[I 2021-08-25 17:40:31.872 DaskGateway] - HTTP routes listening at http://127.0.0.1:59959
[I 2021-08-25 17:40:31.872 DaskGateway] - Scheduler routes listening at gateway://127.0.0.1:59959
[I 2021-08-25 17:40:31.882 DaskGateway] Backend started, clusters will contact api server at http://127.0.0.1:59959/api
[W 2021-08-25 17:40:31.886 Proxy] Unexpected failure fetching routing table, retrying in 0.5s: Get "http://127.0.0.1:42353/api/v1/routes": dial tcp 127.0.0.1:42353: connect: connection refused
[I 2021-08-25 17:40:31.887 DaskGateway] Dask-Gateway server started
[I 2021-08-25 17:40:31.888 DaskGateway] - Private API server listening at http://127.0.0.1:42353
[I 2021-08-25 17:40:32.388 DaskGateway] 200 GET /api/v1/routes 0.570ms
[W 2021-08-25 17:40:32.394 Proxy] Unexpected failure fetching routing events, retrying in 0.5s: Get "http://127.0.0.1:42353/api/v1/routes?watch=1&last_id=1": dial tcp 127.0.0.1:42353: connect: connection refused
[I 2021-08-25 17:40:32.394 DaskGateway] Stopping the Dask gateway proxy
[I 2021-08-25 17:40:32.394 DaskGateway] Stopped successfully
--------------------------- Captured stderr teardown ---------------------------
[E 2021-08-25 17:40:32.405 asyncio base_events:1707] Task was destroyed but it is pending!
    task: <Task pending name='Task-1688' coro=<HTTP1ServerConnection._server_request_loop() running at /opt/hostedtoolcache/Python/3.8.11/x64/lib/python3.8/site-packages/tornado/http1connection.py:823> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7f3c34ea8730>()]> cb=[IOLoop.add_future.<locals>.<lambda>() at /opt/hostedtoolcache/Python/3.8.11/x64/lib/python3.8/site-packages/tornado/ioloop.py:688]>
=============================== warnings summary ===============================
tests/kubernetes/test_helm.py:8
  /home/runner/work/dask-gateway/dask-gateway/tests/kubernetes/test_helm.py:8: PytestUnknownMarkWarning: Unknown pytest.mark.kubernetes - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/mark.html
    pytestmark = pytest.mark.kubernetes

tests/test_client.py::test_client_init
  /opt/hostedtoolcache/Python/3.8.11/x64/lib/python3.8/site-packages/_pytest/unraisableexception.py:78: PytestUnraisableExceptionWarning: Exception ignored in: <function Gateway.__del__ at 0x7f3c438929d0>
  
  Traceback (most recent call last):
    File "/home/runner/work/dask-gateway/dask-gateway/dask-gateway/dask_gateway/client.py", line 376, in __del__
      not self.asynchronous
    File "/home/runner/work/dask-gateway/dask-gateway/dask-gateway/dask_gateway/client.py", line 334, in asynchronous
      return self._asynchronous
  AttributeError: 'Gateway' object has no attribute '_asynchronous'
  
    warnings.warn(pytest.PytestUnraisableExceptionWarning(msg))

-- Docs: https://docs.pytest.org/en/stable/warnings.html
=========================== short test summary info ============================
FAILED tests/test_db_backend.py::test_worker_fails_during_start[1] - ValueErr...
= 1 failed, 110 passed, 7 skipped, 3 deselected, 2 warnings in 65.46s (0:01:05) =
Error: Process completed with exit code 1.
```

--- a/continuous_integration/docker/pbs/files/scripts/start.sh
+++ b/continuous_integration/docker/pbs/files/scripts/start.sh
@@ -32,4 +32,9 @@ echo "PBS_COMM_THREADS=2" >> $PBS_CONF_FILE
 /opt/pbs/bin/qmgr -c "set node pbs queue=workq"
 /opt/pbs/bin/qmgr -c "set server operators += dask@pbs"
 
+# "Entering sleep" can be used as a signal in logs that we have passed the
+# initialization phase where the memory needs may peak and expose us to the
+# OOMKiller and 137 exit codes.
+#
+echo "Entering sleep"
 sleep infinity

--- a/continuous_integration/docker/pbs/files/scripts/start.sh
+++ b/continuous_integration/docker/pbs/files/scripts/start.sh
@@ -6,10 +6,21 @@ MOM_CONF_FILE=/var/spool/pbs/mom_priv/config
 HOSTNAME=$(hostname)
 
 # Configure PBS to run all on one node
+#
+# Configuration references:
+# - https://github.com/openpbs/openpbs/blob/master/doc/man8/pbs.conf.8B
+# - https://github.com/openpbs/openpbs/blob/HEAD/doc/man8/pbs_comm.8B
+# - https://github.com/openpbs/openpbs/blob/master/doc/man8/pbs_mom.8B
+#
 sed -i "s/PBS_SERVER=.*/PBS_SERVER=$HOSTNAME/" $PBS_CONF_FILE
 sed -i "s/PBS_START_MOM=.*/PBS_START_MOM=1/" $PBS_CONF_FILE
 sed -i "s/\$clienthost .*/\$clienthost $HOSTNAME/" $MOM_CONF_FILE
 echo "\$usecp *:/ /" >> $MOM_CONF_FILE
+
+# Reduce the memory footprint by using less threads to avoid the OOMKiller in
+# GitHub Actions as observed with exit code 137.
+#
+echo "PBS_COMM_THREADS=2" >> $PBS_CONF_FILE
 
 # Start PBS
 /etc/init.d/pbs start

--- a/continuous_integration/docker/pbs/start.sh
+++ b/continuous_integration/docker/pbs/start.sh
@@ -14,3 +14,23 @@ docker run --rm -d \
     -p 8088:8088 \
     --cap-add=SYS_RESOURCE \
     ghcr.io/dask/dask-gateway-ci-pbs
+
+# The pbs container's entrypoint, files/scripts/start.sh, emits a log message
+# that we will await observing. We do it to avoid getting OOMKilled by peaking
+# memory needs during startup, which is prone to happen if we run pip install at
+# the same time.
+#
+set +x
+await_startup() {
+    i=0; while [ $i -ne 30 ]; do
+        docker logs pbs 2>/dev/null | grep --silent "Entering sleep" \
+            && start_script_finishing=true && break \
+            || start_script_finishing=false && sleep 1 && i=$((i + 1)) && echo "Waiting for pbs container startup ($i seconds)"
+    done
+    if [ "$start_script_finishing" != "true" ]; then
+        echo "WARNING: /script/start.sh was slow to finish!"
+        exit 1
+    fi
+    echo "pbs container started!"
+}
+await_startup

